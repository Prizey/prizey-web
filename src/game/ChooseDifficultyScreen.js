import React from 'react'
import { withStyles } from '@material-ui/core/styles'
import { Typography } from '@material-ui/core'
import Layout from 'design/Layout/Layout'
import { Redirect } from '@reach/router'

import UserBalance from 'design/UserBalance'
import ProfileLink from 'design/ProfileLink/ProfileLink'
import DifficultyButton from './DifficultyButton'
import SpeedComponent from './SpeedComponent'

const styles = theme => ({
  buttonGroup: {
    marginBottom: theme.spacing.md,
    marginTop: theme.spacing.md,
  },
})

const difficulties = [
  {
    difficulty: 'easy',
    label: 'EASY',
    to: '/game/easy',
  },
  {
    difficulty: 'medium',
    label: 'MEDIUM',
    to: '/game/medium',
  },
  {
    difficulty: 'hard',
    label: 'HARD',
    to: '/game/hard',
  },
]

const ChooseDifficultyScreen = withStyles(styles)(
  ({ navigate, classes, settings, currentUser, location, setCurrentUser }) => (
    <Layout
      location={location}
      currentUser={currentUser}
      leftIcon={<UserBalance />}
      rightIcon={<ProfileLink />}
    >
      <Typography align="center" variant="h5">
        Pick a difficulty!
      </Typography>

      <div className={classes.buttonGroup}>
        {difficulties.map(item => (
          <DifficultyButton
            key={item.difficulty}
            {...item}
            quantity={settings[`${item.difficulty}TicketAmount`]}
            availableTickets={currentUser.tickets}
            navigate={navigate}
            setCurrentUser={setCurrentUser}
          />
        ))}
      </div>

      <Typography align="center">
        The harder difficulty, <br />
        the better the prizes.
      </Typography>
    </Layout>
  ),
)

const RedirectUserWithoutBalance = ({
  navigate,
  currentUser,
  setCurrentUser,
  location,
}) => (
  <SpeedComponent
    render={settings => {
      const userCanPlay = difficulties.filter(
        ({ difficulty }) =>
          currentUser.tickets >= settings[`${difficulty}TicketAmount`],
      )

      return userCanPlay.length ? (
        <ChooseDifficultyScreen
          location={location}
          navigate={navigate}
          settings={settings}
          currentUser={currentUser}
          setCurrentUser={setCurrentUser}
        />
      ) : (
        <Redirect to="/buy-diamonds" noThrow />
      )
    }}
  />
)

export default ({ navigate, setCurrentUser, currentUser, location }) =>
  currentUser ? (
    <RedirectUserWithoutBalance
      navigate={navigate}
      setCurrentUser={setCurrentUser}
      location={location}
      currentUser={currentUser}
    />
  ) : (
    <Redirect to="/sign-in" noThrow />
  )
