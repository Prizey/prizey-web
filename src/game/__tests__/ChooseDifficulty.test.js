import React from 'react'
import renderer from 'react-test-renderer'
import ChooseDifficultyScreen from '../ChooseDifficultyScreen'

const mockList = [
  { id: 1, price: 10.0, ticketAmount: 10 },
  { id: 2, price: 20.0, ticketAmount: 20 },
  { id: 3, price: 30.0, ticketAmount: 30 },
]

jest.mock('../../design/Layout/RegisterPageView')
jest.mock('design/UserBalance', () => () => null)

jest.mock('react-redux', () => ({
  connect: () => Component => props => (
    <Component {...props} insertCoin={jest.fn} />
  ),
}))

jest.mock('croods', () => ({
  List: props => <div {...props}>List: {props.render(mockList)}</div>,
  New: props => (
    <div {...props}>
      New - <div>render - {props.render(props)}</div>
    </div>
  ),
}))

jest.mock('../../design/AdminText/AdminText', () => props => (
  <div {...props}>
    AdminText -{' '}
    {props.render({
      difficultyBottomText: 'The harder difficulty, the better the prizes.',
      difficultyFirstLevelLabel: 'EASY',
      difficultySecondLevelLabel: 'MEDIUM',
      difficultyThirdLevelLabel: 'HARD',
      difficultyTitle: 'Pick a difficulty',
    })}
  </div>
))

jest.mock('@reach/router', () => ({
  Link: ({ children, ...props }) => <div {...props}>Link - {children}</div>,
  Redirect: props => <div {...props}>Redirect</div>,
  navigate: jest.fn(),
}))

it('renders correctly', () => {
  const params = {
    currentUser: {
      tickets: 10,
    },
  }
  const tree = renderer.create(<ChooseDifficultyScreen {...params} />).toJSON()
  expect(tree).toMatchSnapshot()
})

describe("when we don't have the currentUser yet", () => {
  it('redirects correctly', () => {
    const tree = renderer.create(<ChooseDifficultyScreen />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
