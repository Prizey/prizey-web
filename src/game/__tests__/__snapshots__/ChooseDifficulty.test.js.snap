// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  id={1}
  name="game"
  path="/game_setting"
  render={[Function]}
>
  Info -
   
  <div
    className="Component-root-9"
  >
    <div
      className="Component-header-7"
    />
    <div
      className="Component-container-4"
    >
      <img
        alt="Prizey"
        className="Component-logoImage-8"
        src="/logo.png"
      />
      <div
        className="Component-caption-2"
      />
      <div
        className="Component-children-3"
      >
        <h5
          className="MuiTypography-root-10 MuiTypography-h5-26 MuiTypography-alignCenter-33"
        >
          Pick a difficulty!
        </h5>
        <div
          className="Component-buttonGroup-1"
        >
          <div
            afterCreate={[Function]}
            name="tickets"
            path="/ticket_transactions"
            render={[Function]}
          >
            New - 
            <div>
              render - 
              <button
                className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-contained-65 MuiButton-containedPrimary-66 MuiButton-raised-68 MuiButton-raisedPrimary-69 MuiButton-fullWidth-79 Component-root_easy-51"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-55"
                >
                  <div
                    name="products"
                    parentId="easy"
                    path="/products/easy"
                    render={[Function]}
                    renderLoading={[Function]}
                  >
                     - Render: 
                    <img
                      alt="product"
                      className="Component-product-49"
                      src="/mocks/trump-mask.png"
                    />
                    RenderLoading: 
                    <div
                      className="MuiCircularProgress-root-83 MuiCircularProgress-indeterminate-85 Component-loading-48"
                      role="progressbar"
                      style={
                        Object {
                          "height": 30,
                          "width": 30,
                        }
                      }
                    >
                      <svg
                        className="MuiCircularProgress-svg-88"
                        viewBox="22 22 44 44"
                      >
                        <circle
                          className="MuiCircularProgress-circle-89 MuiCircularProgress-circleIndeterminate-91"
                          cx={44}
                          cy={44}
                          fill="none"
                          r={20.2}
                          strokeWidth={3.6}
                          style={Object {}}
                        />
                      </svg>
                    </div>
                  </div>
                  <span
                    className="Component-label-47"
                  >
                    EASY
                  </span>
                  <img
                    alt="diamond"
                    className="Component-icon-46"
                    src="/icons/diamond.png"
                  />
                  <span
                    className="Component-quantity-50"
                  >
                    1
                  </span>
                </span>
                <span
                  className="MuiTouchRipple-root-93"
                />
              </button>
            </div>
          </div>
          <div
            afterCreate={[Function]}
            name="tickets"
            path="/ticket_transactions"
            render={[Function]}
          >
            New - 
            <div>
              render - 
              <button
                className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-contained-65 MuiButton-containedPrimary-66 MuiButton-raised-68 MuiButton-raisedPrimary-69 MuiButton-fullWidth-79 Component-root_medium-53"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-55"
                >
                  <div
                    name="products"
                    parentId="medium"
                    path="/products/medium"
                    render={[Function]}
                    renderLoading={[Function]}
                  >
                     - Render: 
                    <img
                      alt="product"
                      className="Component-product-49"
                      src="/mocks/trump-mask.png"
                    />
                    RenderLoading: 
                    <div
                      className="MuiCircularProgress-root-83 MuiCircularProgress-indeterminate-85 Component-loading-48"
                      role="progressbar"
                      style={
                        Object {
                          "height": 30,
                          "width": 30,
                        }
                      }
                    >
                      <svg
                        className="MuiCircularProgress-svg-88"
                        viewBox="22 22 44 44"
                      >
                        <circle
                          className="MuiCircularProgress-circle-89 MuiCircularProgress-circleIndeterminate-91"
                          cx={44}
                          cy={44}
                          fill="none"
                          r={20.2}
                          strokeWidth={3.6}
                          style={Object {}}
                        />
                      </svg>
                    </div>
                  </div>
                  <span
                    className="Component-label-47"
                  >
                    MEDIUM
                  </span>
                  <img
                    alt="diamond"
                    className="Component-icon-46"
                    src="/icons/diamond.png"
                  />
                  <span
                    className="Component-quantity-50"
                  >
                    3
                  </span>
                </span>
                <span
                  className="MuiTouchRipple-root-93"
                />
              </button>
            </div>
          </div>
          <div
            afterCreate={[Function]}
            name="tickets"
            path="/ticket_transactions"
            render={[Function]}
          >
            New - 
            <div>
              render - 
              <button
                className="MuiButtonBase-root-80 MuiButton-root-54 MuiButton-contained-65 MuiButton-containedPrimary-66 MuiButton-raised-68 MuiButton-raisedPrimary-69 MuiButton-fullWidth-79 Component-root_hard-52"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiButton-label-55"
                >
                  <div
                    name="products"
                    parentId="hard"
                    path="/products/hard"
                    render={[Function]}
                    renderLoading={[Function]}
                  >
                     - Render: 
                    <img
                      alt="product"
                      className="Component-product-49"
                      src="/mocks/trump-mask.png"
                    />
                    RenderLoading: 
                    <div
                      className="MuiCircularProgress-root-83 MuiCircularProgress-indeterminate-85 Component-loading-48"
                      role="progressbar"
                      style={
                        Object {
                          "height": 30,
                          "width": 30,
                        }
                      }
                    >
                      <svg
                        className="MuiCircularProgress-svg-88"
                        viewBox="22 22 44 44"
                      >
                        <circle
                          className="MuiCircularProgress-circle-89 MuiCircularProgress-circleIndeterminate-91"
                          cx={44}
                          cy={44}
                          fill="none"
                          r={20.2}
                          strokeWidth={3.6}
                          style={Object {}}
                        />
                      </svg>
                    </div>
                  </div>
                  <span
                    className="Component-label-47"
                  >
                    HARD
                  </span>
                  <img
                    alt="diamond"
                    className="Component-icon-46"
                    src="/icons/diamond.png"
                  />
                  <span
                    className="Component-quantity-50"
                  />
                </span>
                <span
                  className="MuiTouchRipple-root-93"
                />
              </button>
            </div>
          </div>
        </div>
        <p
          className="MuiTypography-root-10 MuiTypography-body1-19 MuiTypography-alignCenter-33"
        >
          The harder difficulty, 
          <br />
          the better the prizes.
        </p>
      </div>
      <div
        className="Component-footer-5"
      >
        <a
          className="Component-footerLink-6"
          href="https://termly.io/en/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="MuiTypography-root-10 MuiTypography-caption-20 MuiTypography-colorTextSecondary-43 MuiTypography-alignCenter-33"
          >
            TERMS
          </span>
        </a>
        <a
          className="Component-footerLink-6"
          href="https://termly.io/en/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="MuiTypography-root-10 MuiTypography-caption-20 MuiTypography-colorTextSecondary-43 MuiTypography-alignCenter-33"
          >
            PRIVACY
          </span>
        </a>
        <a
          className="Component-footerLink-6"
          href="https://termly.io/en/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="MuiTypography-root-10 MuiTypography-caption-20 MuiTypography-colorTextSecondary-43 MuiTypography-alignCenter-33"
          >
            FAIRNESS
          </span>
        </a>
        <a
          className="Component-footerLink-6"
          href="https://termly.io/en/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="MuiTypography-root-10 MuiTypography-caption-20 MuiTypography-colorTextSecondary-43 MuiTypography-alignCenter-33"
          >
            ALL PRIZES
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`when we don't have the currentUser yet redirects correctly 1`] = `null`;
